File: backend/app/schemas/stock.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class StockSearchItem(BaseModel):
    symbol: str
    name: str | None = None
    price: float | None = None
    change_percent: float | None = None


class StockBasicInfoOut(BaseModel):
    symbol: str
    name: str | None = None
    market: str | None = None
    industry: str | None = None
    listed_date: str | None = None
    total_share: float | None = None
    float_share: float | None = None
    total_market_value: float | None = None
    float_market_value: float | None = None


class StockQuoteOut(BaseModel):
    symbol: str
    name: str | None = None
    price: float | None = None
    change_percent: float | None = None
    change_amount: float | None = None
    volume: float | None = None
    amount: float | None = None
    high: float | None = None
    low: float | None = None
    open: float | None = None
    previous_close: float | None = None
    turnover_rate: float | None = None
    source: str | None = None
    source_timestamp: str | None = None
    timestamp: datetime


class KLineItem(BaseModel):
    trade_date: date
    open: float | None = None
    close: float | None = None
    high: float | None = None
    low: float | None = None
    volume: float | None = None
    amount: float | None = None
    amplitude: float | None = None
    change_percent: float | None = None
    change_amount: float | None = None
    turnover_rate: float | None = None
    ma5: float | None = None
    ma10: float | None = None
    ma20: float | None = None
    macd: float | None = None
    macd_signal: float | None = None
    macd_hist: float | None = None
    rsi: float | None = None


class LatestIndicators(BaseModel):
    ma5: float | None = None
    ma10: float | None = None
    ma20: float | None = None
    macd: float | None = None
    macd_signal: float | None = None
    macd_hist: float | None = None
    rsi: float | None = None


class PredictionOut(BaseModel):
    symbol: str
    close_price: float | None = None
    up_probability: float = Field(ge=0, le=100)
    down_probability: float = Field(ge=0, le=100)
    risk_level: str
    reasons: list[str]
    disclaimer: str
    generated_at: datetime


class PredictionRecordOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    symbol: str
    close_price: float | None = None
    up_probability: float
    down_probability: float
    risk_level: str
    reasons: list[str] | None = None
    disclaimer: str
    created_at: datetime


class StockAnalysisOut(BaseModel):
    basic_info: StockBasicInfoOut
    quote: StockQuoteOut
    latest_indicators: LatestIndicators
    prediction: PredictionOut
    history: list[KLineItem]


class ApiMessage(BaseModel):
    message: str

File: frontend/src/types/stock.ts
export interface StockSearchItem {
  symbol: string
  name?: string | null
  price?: number | null
  change_percent?: number | null
}

export interface StockBasicInfo {
  symbol: string
  name?: string | null
  market?: string | null
  industry?: string | null
  listed_date?: string | null
  total_share?: number | null
  float_share?: number | null
  total_market_value?: number | null
  float_market_value?: number | null
}

export interface StockQuote {
  symbol: string
  name?: string | null
  price?: number | null
  change_percent?: number | null
  change_amount?: number | null
  volume?: number | null
  amount?: number | null
  high?: number | null
  low?: number | null
  open?: number | null
  previous_close?: number | null
  turnover_rate?: number | null
  source?: string | null
  source_timestamp?: string | null
  timestamp: string
}

export interface KLineItem {
  trade_date: string
  open?: number | null
  close?: number | null
  high?: number | null
  low?: number | null
  volume?: number | null
  amount?: number | null
  amplitude?: number | null
  change_percent?: number | null
  change_amount?: number | null
  turnover_rate?: number | null
  ma5?: number | null
  ma10?: number | null
  ma20?: number | null
  macd?: number | null
  macd_signal?: number | null
  macd_hist?: number | null
  rsi?: number | null
}

export interface LatestIndicators {
  ma5?: number | null
  ma10?: number | null
  ma20?: number | null
  macd?: number | null
  macd_signal?: number | null
  macd_hist?: number | null
  rsi?: number | null
}

export interface Prediction {
  symbol: string
  close_price?: number | null
  up_probability: number
  down_probability: number
  risk_level: '低' | '中' | '高' | string
  reasons: string[]
  disclaimer: string
  generated_at: string
}

export interface StockAnalysis {
  basic_info: StockBasicInfo
  quote: StockQuote
  latest_indicators: LatestIndicators
  prediction: Prediction
  history: KLineItem[]
}

export interface StockPredictionRecord extends Omit<Prediction, 'generated_at'> {
  id: number
  created_at: string
}
